### PR TITLE
[OCPCLOUD-1350] Implement machine pre-drain and pre-terminate deletion hooks

### PR DESCRIPTION
**Pull request by JoelSpeed**
This PR implements the feature part of https://github.com/openshift/enhancements/pull/862.

This does not include the conditions part of the enhancement which will follow up in a later PR.

This will ensure that when Machine Deletion Hooks are in play, that the machine controller behaves as described in the enhancement by either blocking the drain or blocking the termination of the Machine.

Something I realised while implementing this is that we should only honour the pre-drain hooks when we are actually attempting a drain. There are certain scenarios where pre-drain hooks wouldn't make sense, eg, when a node doesn't actually exist. So that is included within this PR and can be seen in the unit tests included. 

This is blocked on the [API PR](https://github.com/openshift/api/pull/1050) being merged and the api dependency commit being updated

### CODE REVIEW

**Review by alexander-demichev — APPROVED** [review]
/approve
Looks fine, we can work on merging it after api PR is done

File: pkg/webhooks/v1beta1_suite_test.go [hunk 1]
```diff
@@ -39,6 +40,8 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/envtest"
 )
 
+const alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
```
Comment (alexander-demichev) [inline pkg/webhooks/v1beta1_suite_test.go:43]: just out of curiosity, are we limited by an English alphabet here?

File: pkg/webhooks/machine_webhook_test.go [hunk 2]
```diff
@@ -769,6 +778,45 @@ func TestMachineUpdate(t *testing.T) {
...
+			clusterID:    awsClusterID,
+			baseProviderSpecValue: &kruntime.RawExtension{
+				Object: defaultAWSProviderSpec.DeepCopy(),
+			},
+			baseMachineLifecycleHooks: machinev1.LifecycleHooks{
+				PreDrain: []machinev1.LifecycleHook{preDrainHook},
+			},
+			updateAfterDelete: true,
+			updateMachine: func(m *machinev1.Machine) {
+				m.Annotations = map[string]string{}
```
Comment (alexander-demichev) [inline pkg/webhooks/machine_webhook_test.go:817]: does this update remove the lifecycle hook?

File: pkg/webhooks/v1beta1_suite_test.go (hunk 1 above)
Comment (JoelSpeed) [reply pkg/webhooks/v1beta1_suite_test.go:43]: The validation on the field I had down as `[a-zA-Z]+`, do you think that is too restrictive? The proposal suggests a CamelCase string and something easily computer readable, hence sticking to basic ascii.

**Review by elmiko — COMMENTED** [review]
thanks Joel, makes sense to me. i think it might be nice to have a unit test for `GetChangedLifecycleHooks`, but it's not a blocker for me (i know we are testing it transitively).
/lgtm

**Pull request merged**